### PR TITLE
Proposed changes for ``test_metrics.py``

### PR DESCRIPTION
**Pull request by shantam-8**
The following PR (part of #1351) proposes the following changes for _PredictScorer, _ProbaScorer, and _TresholdScorer as part of the ``test_metrics.py``. Using the guidance present, I have used pytests to make the tests as flexible as possible. Please let me know if I need to add/delete anything else and I would be glad to do the same for the rest of ``test_metrics.py``. Thanks a lot!

### CODE REVIEW

File: test/test_metric/test_metrics.py [hunk 1]
```diff
@@ -326,6 +78,625 @@ def test_sign_flip(self):
...
+
+    scorer = autosklearn.metrics.make_scorer(
+        name=name,
+        score_func=score_func,
+        optimum=optimum,
+        worst_possible_result=worst_possible_result,
+        sign=sign,
+        needs_threshold=True,
+        kwargs={},
+    )
```
Comment (eddiebergman) [inline test/test_metric/test_metrics.py:621]: It seems the signature of this is different from the one in the code, i.e. the `sign` parameter.

https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L242-L253

**Review by eddiebergman — CHANGES_REQUESTED** [review]
Wow, seem really good, thank you :)

`black` does elongate everything quite a bit which is a bit annoying. There are workarounds, i.e. 

```python
(
            np.array([[1.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.0, 1.0]]),
            "roc_auc",
            sklearn.metrics.roc_auc_score,
            *(1, 0, -1, -1.0),
        ),
```

Seeing all these tests like this, it seems like we require a lot of redundant info:
* The `optimum`, `worst_possible_result` and `sign` can all be captured by something like `bounds = (optimum, worst_possible_result)`, where you can infer `sign` from the ordering of whether `optimum > worst_possible_result`
* Looking at how most of this test setup seems to be just reconstructing the metrics that already exist, we could just use the ones [we already have defined](https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L320-L375). It should make the `parametrize`s a good bit shorter and remove the `make_scorer` calls everywhere and make the tests much more readable!

The first part would be for another PR and if you want to work on make that simpler, it would be a nice simple contribution too. The second part would be pretty nice to have for this PR :)

I'll run the tests now!

File: test/test_metric/test_metrics.py (hunk 1 above)
Comment (shantam-8) [reply test/test_metric/test_metrics.py:621]: Thanks a lot for your review! Considering there is no parameter called ``sign`` parameter in ``make_scorer``, I guess I should remove ``sign`` in all the proposed tests? Also, I am not really sure as to how to use the already defined scorers ([the one you mentioned](https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L320-L375>)). For example, while I will be able to use predefined regression scores to create tests for _PredictScorer, I would not be able to make tests for _ProbaScorer and _TresholdScorer.

Comment (eddiebergman) [reply test/test_metric/test_metrics.py:621]: Sorry, I was away from work last week!

Yup, we can remove the `sign` parameter :)

My proposal is to essentially get rid of a lot of boiler plate in the test code, for example, this part in `test_proba_scorer_binary`:

```python
@pytest.mark.parametrize(
    "y_pred, name, score_func, optimum, worst_possible_result, sign, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            "log_loss",
            sklearn.metrics.log_loss,
            0,
            MAXINT,
            1,
            0.0,
        ),
        ...,
        ...,
   ]
)
def test_proba_scorer_binary(
    y_pred: np.array,
    name: str,
    score_func: Callable,
    optimum: float,
    worst_possible_result: float,
    sign: float,
    expected_score: float,
):

    ##########################################
    ### This part here
    ##########################################
    scorer = autosklearn.metrics.make_scorer(
        name=name,
        score_func=score_func,
        optimum=optimum,
        worst_possible_result=worst_possible_result,
        sign=sign,
        needs_proba=True,
        kwargs={},
    )
    ##########################################
```

The reason I think we can get rid of this block is because we basically already have them created in the `metrics/__init__.py` file.

This means it would instead look something like:

```python
@pytest.mark.parametrize(
    "y_pred, score_func, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            sklearn.metrics.log_loss,
            0.0,
        ),
        (
            np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]),
            sklearn.metrics.log_loss,
            1.0986122886681098,
        ),
        (
            np.array([[1.0, 1.0, 1.0], [1.0, 1.0, 1.0], [1.0, 1.0, 1.0]]),
            sklearn.metrics.log_loss,
            -1.0986122886681096,
        ),
    ],
)
def test_proba_scorer_multiclass(
    y_pred: np.ndarray,
    scorer: Scorer,
    expected_score: float,
) -> None:
    """
    Expects
    -------
    * Correct scores generated in case of multilclass data when _ProbaScorer is used.
    """
    y_true = np.array([0, 1, 2])
    result_score = scorer(y_true, y_pred)
    assert expected_score == pytest.approx(result_score)
```

Comment (eddiebergman) [reply test/test_metric/test_metrics.py:621]: Looking further, I think most of the tests for metrics that follow the same thing could really just be condensed into one single test.

```python
@pytest.mark.parametrize(
    "y_pred, y_true, scorer, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            np.array([0, 0, 1, 1]),
            sklearn.metrics.log_loss,
            0.0,
        ),
       ...,
    ],
)
def test_scorer(
    y_pred: np.ndarray,
    y_true: np.ndarray,
    scorer: Scorer,
    expected_score: float,
) -> None:
    result_score = scorer(y_true, y_pred)
    assert expected_score == pytest.approx(result_score)
```

We don't really need to comment the `Expects` thing here because the test would then be very straight forward.

---

For all the tests that wouldn't fit into this set up, we can just leave them as they are I guess.

File: test/test_metric/test_metrics.py [hunk 2]
```diff
@@ -326,6 +75,185 @@ def test_sign_flip(self):
...
+            0.0,
+        ),
+        (
+            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
+            np.array([0, 0, 1, 1]),
+            autosklearn.metrics.log_loss,
+            0.0,
+        ),
+        (
+            np.array([[0.0, 0.0], [0.0, 0.0], [0.0, 0.0], [0.0, 0.0]]),
```
Comment (shantam-8) [inline test/test_metric/test_metrics.py:161]: The workflow test fails due to a sign flip in the case of ``log_loss`` as the ``greater_is_better`` hyperparameter in ``autosklearn.metrics.log_loss`` is set as False [as seen here](https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L388-L395). Hence, to rectify this, should I change ``greater_is_better`` hyperparameter for ``autosklearn.metrics.log_loss`` to True or should I flip the signs in cases where the test fails?

Comment (eddiebergman) [reply test/test_metric/test_metrics.py:161]: I'm not sure which of the cases fail in this particularly test but you can just delete the test cases which rely on the `log_loss` having `greater_is_better=True`, as `log_loss` is something we want to have stay as a minimization metric.

**Comment by eddiebergman** [comment]
Hi @shantam-8,

It looks good to me now, sorry for the deleyad responses, we had some deadlines coming up for other papers that kept us away from auto-sklearn! I pinged @mfeurer to have a quick look

Best,
Eddie

**Review by eddiebergman — APPROVED** [review]

**Review by mfeurer — COMMENTED** [review]
Hey, that looks good. Could you maybe add a comment stating that we test the assembled scorers instead of assembling the scorer objects ourselves in the test?

Besides that, this looks great. Thanks a lot!

**Comment by shantam-8** [comment]
Hello @eddiebergman @mfeurer, sorry to reply so late. I had made a few more updates to ``test_metrics.py``, specifically the ``test_sign_flip`` function (formerly part of ``TestScorer`` class) and the ``test_regression_metrics`` and ``test_classification_metrics`` functions (formerly part of ``TestMetricsDoNotAlterInput`` class). Could you please let me know if these are fine or should I roll them back?
Also, I was going to approach the ``TestMetric`` class. Should I do it as another PR after merging this one as I was not very sure as to how to do it?

**Comment by eddiebergman** [comment]
Hi @shantam-8,

I looked at the `TestMetric` class and I think you can honestly just delete it, given that the metrics all achieve the score they are meant to. It doesn't provide much more over the tests we already have and is just extra noise (imo). @mfeurer do you agree?

The other ones look fine to me.

Edit: I think mfeurer is busy for the nxt few days so deleting that class and if all the tests pass, I think this is ready to merge :)

**Comment by eddiebergman** [comment]
Not sure why the metadata generation tests are failing to be honest. I'll try rerun them

**Comment by eddiebergman** [comment]
Sorry to keep it going, I guess I didn't unfold the `TestMetric` class all the way on github. Can you make sure that every metric we export is tested at least once in the big long test, i.e. that there's one or two cases for each metric of sample input, the true y and the expected score.

**Comment by shantam-8** [comment]
While the long test list does contain metrics like ``accuracy``, ``bal_accuracy``, ``r2``, ``log_loss``, and ``roc_auc``, it misses a huge portion of metrics that are part of the ``autosklearn.metrics.REGRESSION_METRICS`` and ``autosklearn.metrics.CLASSIFICATION_METRICS`` like ``precision``, ``recall``, ``f1``, ``mse``, etcetera. However, as the ``TestMetric``class covered these, would it make sense to maintain it?

**Comment by eddiebergman** [comment]
Sure, we can just keep them in instead of adding them to the huge list then :)

**Comment by eddiebergman** [comment]
Hey @shantam-8,

Very sorry for the slow feedback loop on all of this, we've been pretty busy recently! Many thanks for the contribution and we'll be more responsive for the coming months :)

Best,
Eddie
